server/auth/role.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2020 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissionKeys and
+// limitations under the License.
+//
+
+package auth
+
+import (
+	"encoding/json"
+	"github.com/tikv/pd/pkg/errs"
+	"sort"
+)
+
+const (
+	rolePrefix = "roles"
+)
+
+// Role records role info.
+// Read-Only once created.
+type Role struct {
+	Name        string
+	Permissions map[Permission]struct{}
+}
+
+// jsonRole is used as an intermediate model when marshaling/unmarshaling json data
+// because we need to convert map[Permission]struct{} from/to []Permission first.
+type jsonRole struct {
+	Name        string       `json:"name"`
+	Permissions []Permission `json:"permissions"`
+}
```

Comment:
Merge into one struct.

Reply:
I'm wondering how to merge these structs into one while keeping current behavior unchanged. If we marshal `Role` directly, we will get something like `{"{\"resource\": \"storage\", \"action\": \"get\"}": {}, ...}`.

Reply:
Role may defined as

``` golang
type Role struct {
	Name        string       `json:"name"`
	Permissions []Permission `json:"permissions"`
}
```

Then providing function to simulate the function for `map`.

Reply:
Done.